### PR TITLE
ref(rules): Update apply_rule to enqueue rules with slow conditions

### PR DESCRIPTION
**Pull request by ceorourke**
Update `apply_rule` to add rules with "slow" conditions (aka conditions that query Snuba) to a buffer to be evaluated in a task. If a rule with a slow condition has an `action_match` of "all" it will go to the buffer, and if a rule with an `action_match`" of "any" will go to the buffer if the "fast" conditions have failed.

Closes https://github.com/getsentry/team-core-product-foundations/issues/244

### CODE REVIEW

File: src/sentry/rules/processing/delayed_processing.py [hunk 1]
```diff
@@ -17,9 +21,6 @@
 logger = logging.getLogger("sentry.rules.delayed_processing")
 
 
-PROJECT_ID_BUFFER_LIST_KEY = "project_id_buffer_list"
```
Comment (ceorourke) [inline src/sentry/rules/processing/delayed_processing.py:20]: I just moved this for circular import reasons, it's kinda easier if it's in the processor

File: src/sentry/rules/processing/processor.py [hunk 2]
```diff
@@ -272,9 +299,13 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
 
         state = self.get_state()
         condition_list, filter_list = split_conditions_and_filters(rule.data.get("conditions", ()))
-
-        # Sort `condition_list` so that most expensive conditions run last.
-        condition_list.sort(key=lambda condition: is_condition_slow(condition))
+        slow_conditions = None
+        if features.has("organizations:process-slow-alerts", self.project.organization):
+            fast_conditions, slow_conditions = self.group_conditions_by_speed(condition_list)
+            condition_list = fast_conditions
```
Comment (nhsiehgit) [inline src/sentry/rules/processing/processor.py:305]: 🤔 hmmm can we name this a different variable?

feels a bit smelly using `conditions_list` to derive `fast_conditions`, then re-setting conditions list to _be_ fast_conditions

Comment (nhsiehgit) [reply src/sentry/rules/processing/processor.py:305]: I see so it looks like we're either determining whether to run slow conditions last, or skip slow conditions altogether?

Maybe we can _always_ group conditions by speed, and if they _don't_ have the slow conditions flag, we append the slow conditions to the end of the list?

File: src/sentry/rules/processing/processor.py [hunk 3]
```diff
@@ -272,9 +299,13 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
 
         state = self.get_state()
         condition_list, filter_list = split_conditions_and_filters(rule.data.get("conditions", ()))
-
-        # Sort `condition_list` so that most expensive conditions run last.
-        condition_list.sort(key=lambda condition: is_condition_slow(condition))
+        slow_conditions = None
+        if features.has("organizations:process-slow-alerts", self.project.organization):
```
Comment (nhsiehgit) [inline src/sentry/rules/processing/processor.py:303]: Should we be processing the slow conditions if they have this flag?
Or should we be skipping it?

it looks like we're only processing the fast conditions here?

File: src/sentry/rules/processing/processor.py [hunk 4]
```diff
@@ -272,9 +299,13 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
+            fast_conditions, slow_conditions = self.group_conditions_by_speed(condition_list)
+            condition_list = fast_conditions
+        else:
+            # Sort `condition_list` so that most expensive conditions run last.
+            condition_list.sort(key=lambda condition: is_condition_slow(condition))
 
         for predicate_list, match, name in (
```
Comment (nhsiehgit) [inline src/sentry/rules/processing/processor.py:310]: damn. now i have school house rock stuck in my head 😬

File: src/sentry/rules/processing/processor.py [hunk 5]
```diff
@@ -297,6 +331,11 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
                 )
                 return
 
+        # If we've reached here we have slow conditions to evaluate
+        if slow_conditions:
```
Comment (nhsiehgit) [inline src/sentry/rules/processing/processor.py:335]: Oh i see

IF the org flag is set, then we save the slow conditions to be enqueued rather than run immediately

can we add that as a comment above?

**Review by nhsiehgit — COMMENTED** [review]
few quick comments!

File: src/sentry/rules/processing/processor.py (hunk 3 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:303]: If they don't have the flag they just process them in order of fast to slow like it's done now

**Review by wedamija — COMMENTED** [review]
This looks good to me, just one comment

File: src/sentry/rules/processing/processor.py (hunk 3 above)
Comment (nhsiehgit) [reply src/sentry/rules/processing/processor.py:303]: Yea - the flag makes it sound like we should _only_ be processing them IF they have the flag.

kinda wish it was named better but 🤷

File: src/sentry/rules/processing/processor.py [hunk 6]
```diff
@@ -286,6 +317,9 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
             predicate_func = get_match_function(match)
             if predicate_func:
                 if not predicate_func(predicate_iter):
+                    # prevent return here if slow_conditions exist, because we have more evaluation to do
+                    if slow_conditions:
+                        continue
```
Comment (schew2381) [inline src/sentry/rules/processing/processor.py:322]:
```suggestion
                # prevent return here if slow_conditions exist, because we have more evaluation to do
                if not predicate_func(predicate_iter) and not slow conditions:
```

File: src/sentry/rules/processing/processor.py (hunk 2 above)
Comment (schew2381) [reply src/sentry/rules/processing/processor.py:305]: yeah I think that would be a bit clearer as well cuz in both cases we're sorting by condition speed

Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:318]: not sure why but I couldn't get this to actually work, the condition list looks the same when printed but a test fails 🤔

File: src/sentry/rules/processing/processor.py (hunk 6 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py]: nm I was reading it wrong

File: src/sentry/rules/processing/processor.py [hunk 7]
```diff
@@ -277,9 +314,15 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
-        condition_list.sort(key=lambda condition: is_condition_slow(condition))
+        fast_conditions, slow_conditions = self.group_conditions_by_speed(condition_list)
+        process_slow_conditions_later = features.has(
+            "organizations:process-slow-alerts", self.project.organization
+        )
+        if process_slow_conditions_later:
+            condition_list = fast_conditions
+        else:
+            fast_conditions.extend(slow_conditions)  # type: ignore[arg-type]
+            condition_list = fast_conditions
```
Comment (schew2381) [inline src/sentry/rules/processing/processor.py:325]: nit: I think it might be good to just add the lists and include a comment
```suggestion
            # Condition_list is now ordered by speed
            condition_list = fast_conditions + slow_conditions
```

**Review by schew2381 — DISMISSED** [review]
one small nit but otherwise lgtm!

File: src/sentry/rules/processing/processor.py [hunk 8]
```diff
@@ -240,6 +245,38 @@ def get_state(self) -> EventState:
...
+        if self.event.occurrence_id:
+            return IssueOccurrence.fetch(self.event.occurrence_id, project_id=self.project.id)
+        return None
+
+    def enqueue_rule(self, rule: Rule) -> None:
+        self.buffer = RedisBuffer()
+        self.buffer.push_to_sorted_set(PROJECT_ID_BUFFER_LIST_KEY, rule.project.id)
+
+        occurrence_id = self.get_occurrence()
+        value = json.dumps({"event_id": self.event.event_id, "occurrence_id": occurrence_id})
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:272]: Why do we fetch the occurrence here? Can't we just use the id directly?

File: src/sentry/rules/processing/processor.py [hunk 9]
```diff
@@ -240,6 +245,38 @@ def get_state(self) -> EventState:
...
+    def get_occurrence(self) -> IssueOccurrence | None:
+        if self.event.occurrence_id:
+            return IssueOccurrence.fetch(self.event.occurrence_id, project_id=self.project.id)
+        return None
+
+    def enqueue_rule(self, rule: Rule) -> None:
+        self.buffer = RedisBuffer()
+        self.buffer.push_to_sorted_set(PROJECT_ID_BUFFER_LIST_KEY, rule.project.id)
+
+        occurrence_id = self.get_occurrence()
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:271]: Won't this be an `IssueOccurrence`?

File: src/sentry/rules/processing/processor.py (hunk 7 above)
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:325]: nit: 
```
condition_list = fast_conditions
if not process_slow_conditions_later:
    condition_list = fast_conditions + slow_conditions
```

File: src/sentry/rules/processing/processor.py [hunk 10]
```diff
@@ -290,7 +333,8 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
             predicate_iter = (self.condition_matches(f, state, rule) for f in predicate_list)
             predicate_func = get_match_function(match)
             if predicate_func:
-                if not predicate_func(predicate_iter):
+                if not predicate_func(predicate_iter) and not slow_conditions:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:336]: `and  process_slow_conditions_later and not slow_conditions`? I guess that's kind of handled later too

**Review by wedamija — COMMENTED** [review]
Needs a test that validates the occurrence ids are correct

File: src/sentry/rules/processing/processor.py [hunk 11]
```diff
@@ -290,7 +326,8 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:329]: Reading this again, I think we need to handle it differently for `ALL` vs `ANY`. 

For `ANY`, we want to:
- return early if `not predicate_func(predicate_iter) and not slow_conditions`
- enqueue slow conditions and return if `not predicate_func(predicate_iter) and slow_conditions`
- The rule passes if `predicate_func(predicate_iter)`, so we can ignore slow conditions and never enqueue here.

For `ALL`:
- return early if `not predicate_func(predicate_iter)`. There's no point checking the slow conditions in this case, since we already failed.
- enqueue slow conditions and return if `predicate_func(predicate_iter) and slow_conditions`
- Let the rule pass and not enqueue if `predicate_func(predicate_iter) and not slow_conditions`

I think that we also need to make sure that the `slow_conditions` logic only applies to `condition_list`, and not `filter_list`. If `filter_list` passes, then we should handle conditions and all the slow logic. If it doesn't, we don't need to process conditions.

File: src/sentry/rules/processing/processor.py [hunk 12]
```diff
@@ -302,6 +334,39 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
                 )
                 return
 
+        predicate_func = get_match_function(condition_match)
+        if condition_list:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:338]: I don't think we need this check. `any([]) == False` and `all([]) == True`, so we should just let an empty list go through the same `any/all` logic as a list with items

File: src/sentry/rules/processing/processor.py [hunk 13]
```diff
@@ -302,6 +334,39 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
+                log_string = f"Unsupported condition_match {condition_match} for rule {rule.id}"
+                logger.error(
+                    log_string,
+                    filter_match,
+                    rule.id,
+                    extra={**logging_details},
+                )
+                return
+            predicate_iter = (self.condition_matches(f, state, rule) for f in condition_list)
+            if predicate_iter:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:349]: We should also be able to remove this check

File: src/sentry/rules/processing/processor.py [hunk 14]
```diff
@@ -302,6 +334,39 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
+                        return
+
+                    if slow_conditions and process_slow_conditions_later:
+                        self.enqueue_rule(rule)
+                        return
+
+        if slow_conditions and process_slow_conditions_later:
+            self.enqueue_rule(rule)
+            return
+
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:369]: We should remove this along with the other changes I mentioned

File: src/sentry/rules/processing/processor.py [hunk 15]
```diff
@@ -302,6 +334,39 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
                 )
                 return
 
+        predicate_func = get_match_function(condition_match)
+        if condition_list:
+            if not predicate_func:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:339]: I think we should only error here if `not fast_conditions and not slow_conditions`

Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:339]: isn't this basically `if not condition list`? It's ok if there are no conditions, it shouldn't error

File: src/sentry/rules/processing/processor.py (hunk 12 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:338]: I tried removing this but then it gets stuck on `if condition_match == "any" and not result:` - there isn't a result because there is no condition, and we get a default `condition_match` if one isn't provided.

File: src/sentry/rules/processing/processor.py (hunk 14 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:369]: We need this because of how we define `condition_list`:

```
        condition_list = fast_conditions
        if not process_slow_conditions_later:
            condition_list = fast_conditions + slow_conditions
```

if we have the flag, we need to enqueue the rule by this point

File: src/sentry/rules/processing/processor.py (hunk 15 above)
Comment (wedamija) [reply src/sentry/rules/processing/processor.py:339]: Right, but we don't want to error if there are no conditions. In some cases, we only have `slow_conditions` and no conditions in `condition_list`, so we'd want to error in that case I think?

Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:339]: in that case we just want to go straight to enqueueing the rule, right?

Comment (wedamija) [reply src/sentry/rules/processing/processor.py:339]: If the `predicate_func is None` we'd still want to error if we had `slow_conditions` wouldn't we?

File: src/sentry/rules/processing/processor.py (hunk 12 above)
Comment (wedamija) [reply src/sentry/rules/processing/processor.py:338]: Right, I guess if we have filters and no conditions we don't want to fail on `ANY`...

File: src/sentry/rules/processing/processor.py [hunk 16]
```diff
@@ -277,30 +309,61 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
+                if not result:
+                    return
+
+                if slow_conditions and process_slow_conditions_later:
+                    self.enqueue_rule(rule)
+                    return
+
+        if slow_conditions and process_slow_conditions_later:
+            self.enqueue_rule(rule)
+            return
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:366]: I feel like there are cases where this will get hit where we don't want it to... Should it be in the `else` for `if condition_list`?

Like what if we have 
```
match = "any"
fast_conditions = [c1-pass]
slow_conditions = [c2]
```

So we'll go through the if statement, then `c1-pass` gets a True result so we don't return. But then we'll hit this condition and enqueue, even though the condition has already passed

File: src/sentry/rules/processing/processor.py (hunk 15 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:339]: Yes, I can pull that out to be before the if statement

File: src/sentry/rules/processing/processor.py (hunk 16 above)
Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:367]: Hmm it can't be the else because if we have the flag, slow and fast conditions with an any match, and the fast condition fails, we won't enqueue it. Maybe I need some new boolean that defaults to `False` and gets set to `True` if it's an any match, we have the flag and a fast condition, and then if True we enqueue if there are slow conditions?

File: src/sentry/rules/processing/processor.py [hunk 17]
```diff
@@ -302,6 +334,42 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
                 )
                 return
 
+        predicate_func = get_match_function(condition_match)
+        if not predicate_func:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:338]: We should only error here if there is at least one slow or fast condition present - in the loop, we continued if `predicate_list` was empty.  `not predicate_func and (slow_conditions or fast_conditions)` (We can't use `condition_list` since it might not contain all conditions)

File: src/sentry/rules/processing/processor.py [hunk 18]
```diff
@@ -302,6 +334,42 @@ def apply_rule(self, rule: Rule, status: GroupRuleStatus) -> None:
...
+            log_string = f"Unsupported condition_match {condition_match} for rule {rule.id}"
+            logger.error(
+                log_string,
+                filter_match,
+                rule.id,
+                extra={**logging_details},
+            )
+            return
+        skip_slow_conditions_check = False
+        if condition_list:
```
Comment (wedamija) [inline src/sentry/rules/processing/processor.py:348]: If we replace `condition_list` in this if statement with `fast_conditions or slow_conditions` can we get rid of the if statement outside? This code is getting hard to follow and will be hard to maintain as is, so I'd really like to simplify

Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:348]: You mean get rid of the `if not predicate_func` and logging?

Comment (ceorourke) [reply src/sentry/rules/processing/processor.py:348]: I just added a test for that, we need to `return` `if not predicate_func` otherwise the alert would fire if someone (probably via the API) makes a rule with an unsupported action match (unless the serializer protects against that, but the way this is written it sounds like it doesn't).

Comment (wedamija) [reply src/sentry/rules/processing/processor.py:348]: No, the `if slow_conditions and process_slow_conditions_later and not skip_slow_conditions_check:` after this if statement.

**Review by wedamija — APPROVED** [review]
